datasets/pg19/pg19.py

```diff
@@ -0,0 +1,155 @@
+"""TODO(pg19): Add a description here."""
+
+from __future__ import absolute_import, division, print_function
+
+import csv
+import json
+import os
+
+import tensorflow as tf
+
+import nlp
+
+
+# TODO(pg19): BibTeX citation
+_CITATION = """\
+@article{raecompressive2019,
+  author = {Rae, Jack W and Potapenko, Anna and Jayakumar, Siddhant M and
+            Hillier, Chloe and Lillicrap, Timothy P},
+  title = {Compressive Transformers for Long-Range Sequence Modelling},
+  journal = {arXiv preprint},
+  url = {https://arxiv.org/abs/1911.05507},
+  year = {2019},
+}
+
+"""
+
+# TODO(pg19):
+_DESCRIPTION = """\
+This repository contains the PG-19 language modeling benchmark.
+It includes a set of books extracted from the Project Gutenberg books library, that were published before 1919.
+It also contains metadata of book titles and publication dates.
+"""
+
+_ROOT_URL = "gs://deepmind-gutenberg/"
+_METADATA_URL = os.path.join(_ROOT_URL, "metadata.csv")
+
+
+class Pg19(nlp.GeneratorBasedBuilder):
+    """TODO(pg19): Short description of my dataset."""
+
+    # TODO(pg19): Set up version.
+    VERSION = nlp.Version("0.1.0")
+
+    def _info(self):
+        # TODO(pg19): Specifies the nlp.DatasetInfo object
+        return nlp.DatasetInfo(
+            # This is the description that will appear on the datasets page.
+            description=_DESCRIPTION,
+            # nlp.features.FeatureConnectors
+            features=nlp.Features(
+                {
+                    "short_book_title": nlp.Value("string"),
+                    "publication_date": nlp.Value("int32"),
+                    "url": nlp.Value("string"),
+                    "text": nlp.Value("string"),
+                    # These are the features of your dataset like images, labels ...
+                }
+            ),
+            # If there's a common (input, target) tuple from the features,
+            # specify them here. They'll be used if as_supervised=True in
+            # builder.as_dataset.
+            supervised_keys=None,
+            # Homepage of the dataset for documentation
+            homepage="https://github.com/deepmind/pg19",
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager):
+        """Returns SplitGenerators."""
+        # TODO(pg19): Downloads the data and defines the splits
+        # dl_manager is a nlp.download.DownloadManager that can be used to
+        # download and extract URLs
+
+        def get_filename(path):
+            return os.path.splitext(os.path.basename(path))[0]
+
+        def download_listdir(url, local_filepath):
+            results = tf.io.gfile.listdir(url)
```

Comment:
It would be great if we can make this dataset `tf` independent. Would it be possible to just go with `glob.glob(url)` here?

Reply:
maybe @lhoestq has some idea on this?

Reply:
TF is used here because if allows to browse the google storage as a filesystem to glob and download files.
This is the same for `boolq` if I'm not wrong.

To make it TF-independant we have to 
1) use public urls like `https://storage.googleapis.com/deepmind-gutenberg/train/{i}.txt` instead of `gs://deepmind-gutenberg/train/{i}.txt`, so that we we can use `dl_manager.download_and_extract`.
2) replace the `tf.io.gfile.listdir(url)` with something like `[get_url(i) for i in range(total_number_of_files)]`

Reply:
@lhoestq for #2, we don't know which `i` falls into one of `/train`, `/test`, or `/validation`

Reply:
Ok I see.
You can have the list by calling `https://storage.googleapis.com/storage/v1/b/deepmind-gutenberg/o?prefix=train` (you can use the `requests` package to call it), and take a look at the `name` attribute of each element.

Reply:
@lhoestq very nice! I'll make the change